tfjs-layers/src/layers/nlp/tokenizers.ts

```diff
@@ -436,37 +458,108 @@ export class BytePairTokenizer extends Tokenizer {
-    throw new NotImplementedError(
-      `Not implemented yet. Will depend on ${this.bpeMergeAndUpdateCache},
-       ${this._dtype}, ${this.unicode2Byte}, ${this.tokenToIdMap},
-       ${this.idToTokenMap}`);
+    if (this.addPrefixSpace) {
+      const strInputs = tensorToArr(inputs) as string[];
+      inputs = tensor(strInputs.map(word => ' ' + word));
+    }
+
+    const rawTokensTensor = splitStringsForBpe(inputs, this.unsplittableTokens);
+    const rawTokens = tensorArrTo2DArr(rawTokensTensor) as string[][];
+
+    const tokenRowSplits = [0];
+    for (const [idx, token] of rawTokens.entries()) {
+      tokenRowSplits.push(tokenRowSplits[idx] + token.length);
+    }
+
+    const flatTokens = rawTokens.reduce((acc, e) => acc.concat(e), []);
+
+    // Check cache.
+    const cacheLookup = this.cache.lookup(flatTokens);
+    const cacheMask = cacheLookup.map(e => e === '');
+
+    const emptyFlatTokensMask = flatTokens.map(e => e !== '');
+    const hasUnseenWords = cacheMask.map(
+      (bool, idx) => bool && emptyFlatTokensMask[idx]).some(e => e);
+
+    const processUnseenTokens = (): string[]  => {
+      const unseenTokens = flatTokens.filter((_, idx) => cacheMask[idx]);
+      this.bpeMergeAndUpdateCache(tensor(unseenTokens));
+      return this.cache.lookup(flatTokens);
+    };
+
+    // If `has_unseen_words == True`, it means not all tokens are in cache,
+    // we will process the unseen tokens. Otherwise return the cache lookup.
+    const tokenizedWords =
+      hasUnseenWords ? processUnseenTokens() : cacheLookup;
+
+    const tokensTensor = this.tokenToIdMap.lookup(
+      tokenizedWords.map(word => tensor(word.split(' '))));
+    const tokens = tokensTensor.map(t => Array.from(t.dataSync()));
+
+    // Unflatten to match input.
+    const newTokenRowSplits = [0];
+    for (const [idx, token] of tokens.entries()) {
+      newTokenRowSplits.push(newTokenRowSplits[idx] + token.length);
+    }
+    const newFlatTokens = tokens.reduce((acc, e) => acc.concat(e), []);
+    const gatheredIndices =
+      tokenRowSplits.map(index => newTokenRowSplits[index]);
+
+    const token2D: number[][] = [];
+    for (let i = 0; i < gatheredIndices.length - 1; i++) {
+      const [start, end] = [gatheredIndices[i], gatheredIndices[i+1]];
+      const row = newFlatTokens.slice(start, end);
+      token2D.push(row);
+    }
+
+    // Convert to a dense output if `sequence_length` is set.
+    if (this.sequenceLength) {
+      // pad or truncate
+      const maxLengths = token2D.map(arr => arr.reduce(
+          (a, b) => Math.max(a, b), -Infinity));
+      const maxLength = maxLengths.reduce((a, b) => Math.max(a, b), -Infinity);
+
+      token2D.map(tokenArr => {
+        const pad: number[] = Array(maxLength - tokenArr.length).fill(0);
+
+        const newRow = [];
+        for (let i = 0; i < pad.length; i++) {
+          if (pad[i] === undefined) {
+            break;
+          }
+          newRow.push(tokenArr[i]);
```

Comment:
Ah yes, I was trying to both truncate and pad each array element as needed, but I agree - I don't think this was doing what I thought it was. I've learned how to use `tf.slice` and `tf.pad` since writing this so I've changed this to work directly on tensors now and added coverage tests for this branch.